src/gazebo_mavlink_interface.cpp

```diff
@@ -592,86 +593,128 @@ void GazeboMavlinkInterface::send_mavlink_message(const uint8_t msgid, const voi
+  math::Vector3 pos_g = model_->GetWorldPose().pos; 
+  math::Vector3 pos_n = q_ng.RotateVector(pos_g);
+
+  //gzerr << "got imu: " << C_W_I << "\n";
+  //gzerr << "got pose: " << T_W_I.rot << "\n";
   float declination = get_mag_declination(lat_rad, lon_rad);
 
-  math::Quaternion C_D_I(0.0, 0.0, declination);
+  math::Quaternion q_dn(0.0, 0.0, declination);
+  math::Vector3 mag_n = q_dn.RotateVectorReverse(mag_d_);
 
-  math::Vector3 mag_decl = C_D_I.RotateVectorReverse(mag_W_);
+  math::Vector3 vel_b = q_br.RotateVector(model_->GetRelativeLinearVel());
+  math::Vector3 vel_n = q_ng.RotateVector(model_->GetWorldLinearVel());
+  math::Vector3 omega_nb_b = q_br.RotateVector(model_->GetRelativeAngularVel());
 
-  // TODO replace mag_W_ in the line below with mag_decl
-
-  math::Vector3 mag_I = C_W_I.RotateVectorReverse(mag_decl); // TODO: Add noise based on bais and variance like for imu and gyro
-  math::Vector3 body_vel = C_W_I.RotateVectorReverse(model_->GetWorldLinearVel());
-  
   standard_normal_distribution_ = std::normal_distribution<float>(0, 0.01f);
-
-  float mag_noise = standard_normal_distribution_(random_generator_);
+  math::Vector3 mag_noise_b(
+    standard_normal_distribution_(random_generator_),
+    standard_normal_distribution_(random_generator_),
+    standard_normal_distribution_(random_generator_));
+
+  math::Vector3 accel_b = q_br.RotateVector(math::Vector3(
+    imu_message->linear_acceleration().x(),
+    imu_message->linear_acceleration().y(),
+    imu_message->linear_acceleration().z()));
+  math::Vector3 gyro_b = q_br.RotateVector(math::Vector3(
+    imu_message->angular_velocity().x(),
+    imu_message->angular_velocity().y(),
+    imu_message->angular_velocity().z()));
+  math::Vector3 mag_b = q_nb.RotateVectorReverse(mag_n) + mag_noise_b;
 
   mavlink_hil_sensor_t sensor_msg;
   sensor_msg.time_usec = world_->GetSimTime().nsec*1000;
-  sensor_msg.xacc = imu_message->linear_acceleration().x();
-  sensor_msg.yacc = -imu_message->linear_acceleration().y();
-  sensor_msg.zacc = -imu_message->linear_acceleration().z();
-  sensor_msg.xgyro = imu_message->angular_velocity().x();
-  sensor_msg.ygyro = -imu_message->angular_velocity().y();
-  sensor_msg.zgyro = -imu_message->angular_velocity().z();
-  sensor_msg.xmag = mag_I.x + mag_noise;
-  sensor_msg.ymag = -mag_I.y + mag_noise;
-  sensor_msg.zmag = -mag_I.z + mag_noise;
+  sensor_msg.xacc = accel_b.x;
+  sensor_msg.yacc = accel_b.y;
+  sensor_msg.zacc = accel_b.z;
+  sensor_msg.xgyro = gyro_b.x;
+  sensor_msg.ygyro = gyro_b.y;
+  sensor_msg.zgyro = gyro_b.z;
+  sensor_msg.xmag = mag_b.x;
+  sensor_msg.ymag = mag_b.y;
+  sensor_msg.zmag = mag_b.z;
   sensor_msg.abs_pressure = 0.0;
-  sensor_msg.diff_pressure = 0.5*1.2754*(body_vel.z + body_vel.x)*(body_vel.z + body_vel.x) / 100;
-  sensor_msg.pressure_alt = pos_W_I.z;
+  float q = 1.2754f;
+  // assumed dynamic pressure due to flow aligned with pitot (body x)
+  sensor_msg.diff_pressure = 0.5f*q*vel_b.x*vel_b.x / 100;
```

Comment:
The old formulation used X and Z - this will break tailsitter VTOLs. The right thing would be to project the body x/z vel based on angle of attack.